### PR TITLE
[Merged by Bors] - feat(NumberTheory.FLT.Three): basic properties of a solution to flt3

### PR DESCRIPTION
**Pull request by riccardobrasca**
We add `Solution'` and `Solution`, structures representing solutions to the generalized Fermat's equation `FermatLastTheoremForThreeGen`.  The proof of flt3 will be a descent argument on a given `Solution`.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/NumberTheory/FLT/Three.lean [hunk 1]
```diff
@@ -147,6 +147,60 @@ lemma FermatLastTheoremForThree_of_FermatLastTheoremThreeGen :
...
+  (hcdvd : λ ∣ c)
+  (H : a ^ 3 + b ^ 3 = u * c ^ 3)
+attribute [nolint docBlame] Solution'.a
+attribute [nolint docBlame] Solution'.b
+attribute [nolint docBlame] Solution'.c
+attribute [nolint docBlame] Solution'.u
+
+/-- `Solution` is the same as `Solution'` with the additional assumption that `λ ^ 2 ∣ a + b`. -/
+structure Solution extends Solution' hζ where
+  (hab : λ ^ 2 ∣ a + b)
```
Comment (Ruben-VandeVelde) [inline Mathlib/NumberTheory/FLT/Three.lean:173]: Or these

File: Mathlib/NumberTheory/FLT/Three.lean [hunk 2]
```diff
@@ -147,6 +147,60 @@ lemma FermatLastTheoremForThree_of_FermatLastTheoremThreeGen :
...
+variable {hζ} (S : Solution hζ) (S' : Solution' hζ) [DecidableRel fun (a b : 𝓞 K) ↦ a ∣ b]
+
+/-- For any `S' : Solution'`, the multiplicity of `λ` in `S'.c` is finite. -/
+lemma Solution'.multiplicity_lambda_c_finite :
+    multiplicity.Finite (hζ.toInteger - 1) S'.c :=
+  multiplicity.finite_of_not_isUnit hζ.zeta_sub_one_prime'.not_unit S'.hc
+
+/-- Given `S' : Solution'`, `S'.multiplicity` is the multiplicity of `λ` in `S'.c`, as a natural
+number. -/
+def Solution'.multiplicity  :=
```
Comment (Ruben-VandeVelde) [inline Mathlib/NumberTheory/FLT/Three.lean:184]:
```suggestion
def Solution'.multiplicity :=
```

**Review by Ruben-VandeVelde — APPROVED** [review]

File: Mathlib/NumberTheory/FLT/Three.lean [hunk 3]
```diff
@@ -147,6 +147,60 @@ lemma FermatLastTheoremForThree_of_FermatLastTheoremThreeGen :
...
+number. -/
+def Solution'.multiplicity :=
+  (_root_.multiplicity (hζ.toInteger - 1) S'.c).get (multiplicity_lambda_c_finite S')
+
+/-- Given `S : Solution`, `S.multiplicity` is the multiplicity of `λ` in `S.c`, as a natural
+number. -/
+def Solution.multiplicity := S.toSolution'.multiplicity
+
+/-- We say that `S : Solution` is minimal if for all `S₁ : Solution`, the multiplicity of `λ` in
+`S.c` is less or equal than the multiplicity in `S'.c`. -/
```
Comment (kbuzzard) [inline Mathlib/NumberTheory/FLT/Three.lean:192]:
```suggestion
`S.c` is less or equal than the multiplicity in `S₁.c`. -/
```

**Review by kbuzzard — APPROVED** [review]
This all looks fine to me. Why are we formalising this proof rather than the one over integers in FLT-regular, by the way?

bors d+

**Comment by riccardobrasca** [comment]
> This all looks fine to me. Why are we formalising this proof rather than the one over integers in FLT-regular, by the way?
> 
> bors d+

No particular reasons, it seemed a good excuse to develop the relevant API and test the current one.

**Pull request closed without merging**